text/0000-ok-coercion.md

```diff
@@ -0,0 +1,578 @@
+Rather than `panic!` on an error, this will return the error value from the current function for the caller to handle.
+
+**<ins>This function should return</ins>** an `Ok` value in the success case. We've declared the `run` function's
+success type as `()` in the signature, which means we **<ins>should</ins>** wrap the unit type value in the `Ok` value.
+**<ins>We could explicitly return a value of `Ok(())`, but this syntax looks strange and does not add much value to the
+code, thus the compiler allows us to leave it off when the result is exactly `Ok(())`.</ins>**
+
+When you run this, it will compile, but with a warning:
+
+```text
+warning: unused result which must be used, #[warn(unused_must_use)] on by default
+  --> src/main.rs:39:5
+   |
+39 |     run(config);
+   |     ^^^^^^^^^^^^
+```
+
+Rust is telling us that our code ignores the `Result` value, which might be indicating that there was an error. We're
+not checking to see if there was an error or not, though, and the compiler is reminding us that we probably meant to
+have some error handling code here! Let's rectify that now.
+
+[book-12-03]: https://doc.rust-lang.org/book/second-edition/ch12-03-improving-error-handling-and-modularity.html#handling-errors-returned-from-run-in-main
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Coercion
+
+We amend [RFC 401] to add a coercion rule from `()` to `T` if `T` implements `Try<Ok = ()>`. The coerced value will be
+created as `Try::from_ok(())`.
+
+Since coercion site happens outside of the end of a block, this means the following statements will now become valid:
+
+```rust
+// let is a coercion site:
+let _: Result<()> = ();
+
+// array literal is a coercion site:
+let _: Result<()> = [Ok(()), (), Err(e)];
+let _: Result<()> = [(); 3];
+
+// tuple is a coercion site:
+let _: (Result<()>, Result<()>) = ((), ());
+
+// box expression is a coercion site:
+let _: Box<Result<()>> = box ();
+let _: Box<Result<()>> = HEAP <- ();
+```
+
+While coercion is transitive, all transitive rules are irrelevant to `()` → `impl Try`:
+
+| Rule | Why irrelevant |
+|------|----------------|
+| subtyping | `()` has no lifetime, `()` can only be the subtype of itself |
+| `&mut T` → `&T` | `Try` is not implemented for mutable references |
+| `*mut T` → `*const T` | `Try` is not implemented for mutable pointers |
+| `&T` → `*const T` | `Try` is not implemented for shared references |
+| `&mut T` → `*mut T` | `Try` is not implemented for mutable references |
+| closure → `fn` | `Try` is not implemented for closures |
+| CoerceUnsized | `CoerceUnsized` is not implemented for `Result`/`Option` |
+| `()` → `impl Try` | `Try` is not implemented for `()` |
+
+It should be explicitly mentioned that the `()` → `impl Try` rule does not participate in "receiver coercion" (similarly
+the closure → `fn` also does not participate), to avoid truly nonsense expressions like:
```

Comment:
I'm not convinced that `Result::unwrap_err(())` is any less "nonsense" than `().unwrap_err()`.

And [`?` desugars](https://github.com/rust-lang/rfcs/blob/master/text/1859-try-trait.md#desugaring-and-the-try-trait) to non-method form, so removing receiver coercion doesn't impact that.